code/test/security.js

```diff
@@ -0,0 +1,214 @@
+  let revealParams
+  // Process the Operation
+  switch (operationType) {
+    // Format commit and revealParams for TRANSFER Tranasction
+    case ONEConstants.OperationType.TRANSFER:
+      paramsHash = ONEWallet.computeTransferHash
+      commitParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      revealParams = { operationType, tokenType, contractAddress, tokenId, dest, amount, data }
+      break
+    case ONEConstants.OperationType.SET_RECOVERY_ADDRESS:
+      paramsHash = ONEWallet.computeDestOnlyHash
+      commitParams = { operationType, dest }
+      revealParams = { operationType, dest }
+      break
+    case ONEConstants.OperationType.CHANGE_SPENDING_LIMIT:
+    case ONEConstants.OperationType.JUMP_SPENDING_LIMIT:
+      paramsHash = ONEWallet.computeAmountHash
+      commitParams = { operationType, amount }
+      revealParams = { operationType, amount }
+      break
+    default:
+      console.log(`Invalid Operation passed`)
+      assert.strictEqual('A Valid Operation', operationType, 'Error invalid operationType passed')
+      return
+  }
+  let { tx, authParams, revealParams: returnedRevealParams } = await TestUtil.commitReveal({
+    Debugger,
+    layers: walletInfo.layers,
+    index,
+    eotp,
+    paramsHash,
+    commitParams,
+    revealParams,
+    wallet: walletInfo.wallet
+  })
+  let currentState
+  if (getCurrentState) { currentState = await TestUtil.getState(walletInfo.wallet) }
+  return { tx, authParams, revealParams: returnedRevealParams, currentState }
+}
+
+contract('ONEWallet', (accounts) => {
+  Logger.debug(`Testing with ${accounts.length} accounts`)
+  Logger.debug(accounts)
+  let snapshotId
+  beforeEach(async function () {
+    await TestUtil.init()
+    snapshotId = await TestUtil.snapshot()
+  })
+  afterEach(async function () {
+    await TestUtil.revert(snapshotId)
+  })
+
+  // === BASIC POSITIVE TESTING SECURITY ====
+
+  // ==== DISPLACE =====
+  // Test must allow displace operation using 6x6 otps for different durations
+  // Expected result: can authenticate otp from new core after displacement
+  // Note: this is currently tested in innerCores.js
+  it('CO-BASIC-7 DISPLACE : must allow displace operation using 6x6 otps for different durations authenticate otp from new core after displacement', async () => {
+    assert.strictEqual(0, 1, 'Please test displace using innerCores.js')
+  })
+
+  // ====== CHANGE_SPENDING_LIMIT ======
+  // Test changing the spending limit
+  // Expected result alice spending limit will be updated
+  // Change Logic:
+  // Too early: Can't increase the limit twice within the same interval (ss.lastLimitAdjustmentTime + ss.spendingInterval > block.timestamp && newLimit > ss.spendingLimit)
+  // Too much : Can't increase the limit by more than double existing limit + 1 native Token (newLimit > (ss.spendingLimit) * 2 + (1 ether))
+  it('CO-BASIC-24 CHANGE_SPENDING_LIMIT: must be able to transfer native currency', async () => {
```

Comment:
need to test effectiveness of constraints on time. See spendLimit.js, or use the client UI and see the explanations there

Reply:
I have replicated spendingLimt into security.js 🙏 
Two items
a) Do we need to test`timeOffset` when calling `getEOTP`
b) Need to improve event validation for multiple events and negative events (tracked https://github.com/polymorpher/one-wallet/issues/207)